tools/inc/stljobs.h

```diff
@@ -0,0 +1,606 @@
+            }
+
+            readIo.wait(false);
+        }
+    }
+
+    [[nodiscard]] std::string extract() {
+        stop();
+        std::string result(targetBuffer.data(), validTill);
+        start();
+        return result;
+    }
+
+    [[nodiscard]] std::string extract_and_reset() {
+        stop();
+        auto first = targetBuffer.data();
+        auto last  = first + validTill;
+        first      = std::find_if_not(first, last, is_space);
+        last       = std::find_if_not(std::reverse_iterator(last), std::reverse_iterator(first), is_space).base();
+        std::string result(first, static_cast<size_t>(last - first));
+        validTill = 0;
+        start();
+        return result;
+    }
+
+    [[nodiscard]] const std::string& get_buffer() const noexcept {
+        assert(!running.load());
+        return targetBuffer;
+    }
+
+    [[nodiscard]] HANDLE get_write_pipe() noexcept {
+        return writeHandle.get();
+    }
+
+private:
+    static void __stdcall callback(
+        PTP_CALLBACK_INSTANCE, void* thisRaw, void*, ULONG ioResult, ULONG_PTR bytes, PTP_IO) noexcept {
+        switch (ioResult) {
+        case ERROR_SUCCESS:
+        case ERROR_OPERATION_ABORTED:
+            break;
+        default:
+            api_failure("StartThreadpoolIo callback", ioResult);
+            break;
+        }
+
+        const auto this_ = static_cast<output_collecting_pipe*>(thisRaw);
+        this_->validTill += bytes;
+        if (this_->running) {
+            this_->read_some();
+        }
+    }
+
+    void resize_create_target_buffer() {
+        if (bufferSize < targetBuffer.size() - validTill) {
+            // already has enough space
+            return;
+        }
+
+        targetBuffer.resize(validTill + bufferSize);
+        targetBuffer.resize(std::min(static_cast<size_t>(ULONG_MAX), targetBuffer.capacity()));
+    }
+
+    void read_some() {
+        assert(running.load());
+        DWORD bytesRead = 0;
+        readIo.start_threadpool_io();
+        while (resize_create_target_buffer(),
+            ReadFile(readIo.get_file(), targetBuffer.data() + validTill,
+                static_cast<unsigned long>(targetBuffer.size() - validTill), &bytesRead, &overlapped)) {
+            validTill += bytesRead;
+        }
+
+        const auto lastError = GetLastError();
+        if (lastError == ERROR_IO_PENDING) {
+            return;
+        }
+
+        readIo.cancel_threadpool_io();
+        api_failure("ReadFile", lastError);
+    }
+
+    std::atomic<bool> running{};
+    std::string targetBuffer; // if running, owned by a threadpool thread, otherwise owned by the calling thread
+    std::size_t validTill{};
+    handle<invalid_handle_value_policy> writeHandle;
+    tp_io readIo;
+    OVERLAPPED overlapped{};
+};
+
+struct execution_result {
+    unsigned long exitCode;
+    std::string output;
+};
+
+class environment_block {
+public:
+    environment_block() {
+        default_block defaultBlock;

```

Comment:
It would be helpful to have a comment here clarifying that we are trying to find the size of the environment block which is a zero separated sequence of strings ending with two zeroes. Hence, needing to count first before assigning it to std::string.